### PR TITLE
[mlir][emitc] Arith to EmitC: Handle addi, subi and muli

### PR DESCRIPTION
**Pull request by mgehre-amd**
Important to consider that `arith` has wrap around semantics, and in C++ signed overflow is UB.
Unless the operation guarantees that no signed overflow happens, we will perform the arithmetic in an equivalent unsigned type.

### CODE REVIEW

**Comment by marbre** [comment]
> No handling yet for divsi and divui, as they require special considerations for signedness.

Probably we already need some special considerations for this case. E.g. for `arith.addi` the docs https://mlir.llvm.org/docs/Dialects/ArithOps/#arithaddi-arithaddiop state

> Performs N-bit addition on the operands. The operands are interpreted as unsigned bitvectors. The result is represented by a bitvector containing the mathematical value of the addition modulo 2^n, where n is the bitwidth. 

Following https://en.cppreference.com/w/cpp/language/operator_arithmetic#Overflows

> Unsigned integer arithmetic is always performed modulo 2n where n is the number of bits in that particular integer.

we should probably cast to unsigned, perform the arithmetic operation and add another cast. WDYT?

**Review by marbre — COMMENTED** [review]
> No handling yet for divsi and divui, as they require special considerations for signedness.

Probably we already need some special considerations for this case. E.g. for `arith.addi` the docs https://mlir.llvm.org/docs/Dialects/ArithOps/#arithaddi-arithaddiop state

> Performs N-bit addition on the operands. The operands are interpreted as unsigned bitvectors. The result is represented by a bitvector containing the mathematical value of the addition modulo 2^n, where n is the bitwidth.

Following https://en.cppreference.com/w/cpp/language/operator_arithmetic#Overflows

> Unsigned integer arithmetic is always performed modulo 2n where n is the number of bits in that particular integer.

we should probably cast to unsigned, perform the arithmetic operation and add another cast. WDYT?

**Comment by mgehre-amd** [comment]
> > No handling yet for divsi and divui, as they require special considerations for signedness.
> 
> Probably we already need some special considerations for this case. E.g. for `arith.addi` the docs [mlir.llvm.org/docs/Dialects/ArithOps/#arithaddi-arithaddiop](https://mlir.llvm.org/docs/Dialects/ArithOps/#arithaddi-arithaddiop) state
> 
> > Performs N-bit addition on the operands. The operands are interpreted as unsigned bitvectors. The result is represented by a bitvector containing the mathematical value of the addition modulo 2^n, where n is the bitwidth.
> 
> Following [en.cppreference.com/w/cpp/language/operator_arithmetic#Overflows](https://en.cppreference.com/w/cpp/language/operator_arithmetic#Overflows)
> 
> > Unsigned integer arithmetic is always performed modulo 2n where n is the number of bits in that particular integer.
> 
> we should probably cast to unsigned, perform the arithmetic operation and add another cast. WDYT?

Good point, this will otherwise cause UB for when the C types are signed.

File: mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp [hunk 1]
```diff
@@ -55,6 +55,50 @@ class ArithOpConversion final : public OpConversionPattern<ArithOp> {
...
+
+  LogicalResult
+  matchAndRewrite(ArithOp op, typename ArithOp::Adaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+
+    Type type = this->getTypeConverter()->convertType(op.getType());
+    if (!isa_and_nonnull<IntegerType, IndexType>(type)) {
+      return rewriter.notifyMatchFailure(op, "expected integer type");
+    }
+
```
Comment (simon-camp) [inline mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:71]: Return a match failure for `i1`, as these are emitted as `bool` and the arithmetic doesn't wrap around as expected. We can add additional patterns in a follow up.

File: mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir [hunk 2]
```diff
@@ -37,6 +37,57 @@ func.func @arith_ops(%arg0: f32, %arg1: f32) {
 
 // -----
 
+// CHECK-LABEL: arith_integer_ops
+func.func @arith_integer_ops(%arg0: i32, %arg1: i32) {
+  // CHECK: %[[C1:[^ ]*]] = emitc.cast %arg0 : i32 to ui3
```
Comment (simon-camp) [inline mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir:42]:
```suggestion
  // CHECK: %[[C1:[^ ]*]] = emitc.cast %arg0 : i32 to ui32
```
Same for the lines below

**Review by simon-camp — CHANGES_REQUESTED** [review]

**Review by marbre — APPROVED** [review]
Thanks, looks good now :) Just one minor _wish_, could you maybe add a more descriptive commit message? I think what is in the PR description right now is more additional context for review but not necessarily adding value as part of a commit message.

**Pull request merged**